app/models/shipit/deploy_spec/npm_discovery.rb

```diff
@@ -79,7 +93,73 @@ def package_version
         JSON.parse(package_json.read)['version']
       end
 
+      def publish_config
+        JSON.parse(package_json.read)['publishConfig']
+      end
+
+      def publish_config_access
+        publish_config['access']
+      end
+
+      def publish_config_registry
+        publish_config['registry'] || publish_config['@shopify:registry']
+      end
+
+      def package_name
+        JSON.parse(package_json.read)['name']
+      end
+
+      def scoped_package?
+        package_name.start_with?('@shopify')
+      end
+
+      def publish?
+        return false if publish_config.blank?
+        return false unless valid_publish_config_access?
+        return false unless valid_publish_config_registry?
+      end
+
+      def valid_publish_config_access?
+        return false if publish_config_access.blank?
+        return false unless VALID_ACCESS.include?(publish_config_access)
+        true
+      end
+
+      def valid_publish_config_registry?
+        return false if publish_config_registry.blank?
+        return false unless VALID_REGISTRY.include?(publish_config_registry)
```

Comment:
Is this possible?
- Don't require registry info
- If registry is omitted && access === 'public': 
  - `npm publish --access public --registry https://registry.npmjs.org/`
- If registry is omitted && access === 'restricted': 
  - `npm publish --access restricted --registry https://packages.shopify.io/shopify/node/npm/`

Reply:
I don't see registry as a documented option here: https://docs.npmjs.com/cli/publish
Even if it works, we don't have a way to deploy scoped registry (`@shopify:registry`) via CLI - or at least I haven't come across one.
Since the global npmrc will point @shopify to pacakgeCloud, we can change that to the public npm registry via CLI.

Reply:
I recall testing this out via CLI and it didn't work. Lemme know if it happens to work for you.

Reply:
It's documented here: https://docs.npmjs.com/misc/registry#can-i-run-my-own-private-registry

> If you then want to publish a package for the whole world to see, you can simply override the --registry option for that publish command.

Reply:
If the global .npmrc is a problem, create one on the fly inside the checked out folder.

Reply:
It all works, great suggestion! 

This is what I've tested locally: 
* My global `~/.npmrc` had both a write token to PC and write token to our shopify scoped npm account. The tests worked as expected
```
always-auth=true
@shopify:registry=https://packages.shopify.io/shopify/node/npm/
//packages.shopify.io/shopify/node/npm/:_authToken=<PC_TOKEN>
//registry.npmjs.org/:_authToken=<NPM_TOKEN>
```

### Publishes to *public* registry with local .npmrc 
- [x] with  `@shopify` scope:
```
always-auth=true
@shopify:registry=https://registry.npmjs.org/
```

- [x] without scoped name:

```
always-auth=true
registry=https://registry.npmjs.org/
```
	
### Publishes to *private* registry with local .npmrc
- [x] with `@shopify` scope :

```
always-auth=true
@shopify:registry=https://packages.shopify.io/shopify/node/npm/
```

Reply:
PR has been updated to reflect what we've discussed above.